### PR TITLE
[PLAT-2110] Use deamon threads to start flushing sessions

### PR DESCRIPTION
**Pull request by Pezzah**
## Goal

In some cases the Bugsnag notifier can prevent clean application shutdowns due to the session flushing threads that are running.

This changes the threads to be marked as "deamon" so that the application can terminate without waiting for them to finish

The deamon threads run every 60 seconds to flush sessions created during that time. New threads are used to actually perform the flushing which are not marked as deamon so that they won't get killed part way though sending data.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A version of {@link Executors#defaultThreadFactory()} which returns deamon threads
+ * This is to prevent applications from hanging waiting for the sessions scheduled task
+ * because deamon threads will be terminated on application shutdown
+ */
+public class DeamonThreadFactory implements ThreadFactory {
```
Comment (fractalwrench) [inline bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: The implementation of this class could be simplified by using the decorator pattern. We could create a field for `Executors.defaultThreadFactory()`, then make the thread returned by `threadFactory.newThread` a daemon before returning. That might also make a good candidate for a unit test.

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        namePrefix = "deamon-pool-"
+                + poolNumber.getAndIncrement()
+                + "-thread-";
+    }
+
+    /**
+     * @see ThreadFactory#newThread(Runnable)
+     */
+    public Thread newThread(Runnable runner) {
```
Comment (fractalwrench) [inline bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: This is missing an override annotation

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+public class DeamonThreadFactory implements ThreadFactory {
+    private static final AtomicInteger poolNumber = new AtomicInteger(1);
+    private final ThreadGroup group;
+    private final AtomicInteger threadNumber = new AtomicInteger(1);
+    private final String namePrefix;
+
+    /**
+     * Constructor
+     */
+    public DeamonThreadFactory() {
```
Comment (fractalwrench) [inline bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: There's a typo in the class name (Deamon -> Daemon)

File: bugsnag/src/main/java/com/bugsnag/Bugsnag.java [hunk 4]
```diff
@@ -81,7 +84,18 @@ private void scheduleSessionFlushes() {
         sessionExecutorService.scheduleAtFixedRate(new Runnable() {
             @Override
             public void run() {
-                sessionTracker.flushSessions(new Date());
+                // Create a new thread which is not a deamon thread to actually flush the sessions
+                // This will means that the thread will not be killed
+                // half way through if the application shuts down
+                Thread nonDeamonThread = new Thread(new Runnable() {
```
Comment (fractalwrench) [inline bugsnag/src/main/java/com/bugsnag/Bugsnag.java]: We could use a cached executor e.g. `Executors.newSingleThreadExecutor()` rather than creating a new thread every time this is invoked.

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java (hunk 1 above)
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: changed

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java (hunk 2 above)
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: Added

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java (hunk 3 above)
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: fixed

File: bugsnag/src/main/java/com/bugsnag/Bugsnag.java (hunk 4 above)
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/Bugsnag.java]: Added new newSingleThreadExecutor

File: bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java (hunk 1 above)
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/util/DeamonThreadFactory.java]: also added a test

**Comment by kattrali** [comment]
This issue came up before with a single thread executor in #50 ([current logic on master](https://github.com/bugsnag/bugsnag-java/blob/1ff0be9af090b0e4913160249c0952a1ff934d31/bugsnag/src/main/java/com/bugsnag/delivery/AsyncHttpDelivery.java#L21-L26)) - should the session flushing logic use an executor service in the same way to ensure the service doesn't hang an app? I'd just like to check that we aren't introducing a different way for the application to hang by fixing this one.
